feat(treesitter): add ability to reload parser

DanilaMihailov opened the pull request:
Problem:
There is no way to reload current parser for buffer. This is needed when
`vim.opt.rtp` changes and new paths have treesitter queries. For example
when loading plugins lazily.

Solution:
Add options `reload` to `vim.treesitter.get_parser()`, that will
remove previous parser and create new one, effectively loading new
queries from `runtimepath`

clason commented:
Do we really need a new method for this? Can't we instead add a `reload` or `force` parameter (name TBD) to `vim.treesitter.query.get()`? 

(Also needs tests.)

DanilaMihailov commented:
> Do we really need a new method for this? Can't we instead add a `reload` or `force` parameter (name TBD) to `vim.treesitter.query.get()`?
> 
> (Also needs tests.)

Not sure about `vim.treesitter.query.get()`. The problem is I want to update `LanguageTree.__injection_query` field in parser, but it is a private field. Maybe `vim.treesitter.get_parser(bufnr, lang, {force = true})`? That would remove current parser and create new one with `vim.treesitter._create_parser()` and effectively reload `_injection_query`.

I would love to add tests, but is there an example how to test function that depends on `vim.opt.rtp` changes? As I understand I would need to create parser, than add new paths to `vim.opt.rtp` and then reload quieries. Should I store queries somewhere or just create files on the fly and then remove?

DanilaMihailov commented:
Here is an example how I would use this method

```lua
-- after plugin is loaded check if there is any open lua buffers,
-- if so, we need to refresh their parsers, so our treesitter queries will work
for bufnr in vim.iter(vim.api.nvim_list_bufs()) do
  if vim.api.nvim_buf_is_loaded(bufnr) and vim.bo[bufnr].ft == "lua" then
    local parser = vim.treesitter.get_parser(bufnr)
    parser:load_injection_query()
    parser:parse(true)
  end
end
```

clason commented:
> Maybe vim.treesitter.get_parser(bufnr, lang, {force = true})? That would remove current parser and create new one with vim.treesitter._create_parser() and effectively reload _injection_query.

Yes, that is exactly what I was thinking of. 

> I would love to add tests, but is there an example how to test function that depends on vim.opt.rtp changes? As I understand I would need to create parser, than add new paths to vim.opt.rtp and then reload quieries. Should I store queries somewhere or just create files on the fly and then remove?

Neovim already ships with queries in runtimepath, so, yes, creating a temporary file and adding that directory to runtimepath could work. The best thing would be to look through the `test/functional/treesitter` tests and see if there is one that loads queries from a file, then adapt it.

DanilaMihailov commented:
> Yes, that is exactly what I was thinking of. What I don't like about it is that it's specific to injections, but you (sometimes) want to reload the main queries. Even if that is the actual effect, the API is misleading then, hurting discoverability.

Yeah, more general "reload parser" is what I was looking for initially. So I would do it this way. But there is a lot happening with callbacks, that I don't fully grasp. Do I need to do any specific clean up for this:

```lua
  api.nvim_buf_attach(
    source,
    false,
    { on_bytes = bytes_cb, on_detach = detach_cb, on_reload = reload_cb, preview = true }
  )
```

clason commented:
I don't think so; tests will show.

DanilaMihailov committed:
feat(treesitter): add ability to reload parser

Problem:
There is no way to reload current parser for buffer. This is needed when
`vim.opt.rtp` changes and new paths have treesitter queries. For example
when loading plugins lazily.

Solution:
Add options `reload` to `vim.treesitter.create_parser()`, that will
remove previous parser and create new one, effectively loading new
queries from `runtimepath`

justinmk commented on runtime/lua/vim/treesitter.lua:
```diff
@@ -106,6 +106,16 @@ function M.get_parser(bufnr, lang, opts)
+  elseif opts.reload then
+    opts.reload = nil -- do not pass reload to _create_parser
+    assert(lang, 'lang should be valid')
+    parsers[bufnr]:destroy()
+    parsers[bufnr] = M._create_parser(bufnr, lang, opts)
+
+    if M.highlighter.active[bufnr] then
+      M.highlighter.active[bufnr]:destroy()
+      M.highlighter.new(parsers[bufnr])
+    end
```
why is this different than `LanguageTree:invalidate()` ? in fact we have something called `_on_reload` that currently calls invalidate(): https://github.com/neovim/neovim/blob/c1396afa7c41c2009d06a54c8aaf1ca5689a9786/runtime/lua/vim/treesitter/languagetree.lua#L1042-L1044

Would be very helpful to determine which semantics are correct and tighten things up

DanilaMihailov replied:
`invalidate()` does not reload `LanguageTree._inject_query` which is the reason I want to reload the parser. If there is different way it should be done, I am happy to do it.

What happens there is basically what is going to happen if `:edit` executed. When `:edit` executed fires event "detach" for buffer, which triggers `detach_cb` and this deletes parser. Then `vim.treesitter.start()` called by ft plugin and creates new parser.

https://github.com/neovim/neovim/blob/4e5c633ed4871a948aff7338b793ac5f93484153/runtime/lua/vim/treesitter.lua#L47-L52

Right now my goal is to make parser read queries from filesystem again, because I change `vim.opt.rtp` where I add more queries from plugin.

DanilaMihailov replied:
I was thinking, maybe a better idea would be refreshing queries in `invalidate`? Something like this
```lua
function LanguageTree:invalidate(reload, refresh_queries)
  self._valid = false

  if refresh_queries then
    if not self._opts.injections then
      query._reset_cache(self:lang(), 'injections')
      self._injection_query = query.get(self:lang(), 'injections')
    end
    self:_do_callback('changedquery', self:lang())
  end
    -- ...
```

and then highlighter.lua would register callback

```lua
 -- ...
    on_changedquery = function(lang)
      -- just remove queires, they are loaded later on demand
      query._reset_cache(lang, 'highlights')
      self._queries[lang] = nil
    end,
```

lewis6991 replied:
`invalidate` is only supposed to be called when the tracked state of the Languagetree is not valid against the parse tree in tree-sitter, and is called fairly often. I expect clearing file system caches will kill performance.

justinmk replied:
> `invalidate` is only supposed to be called when the tracked state of the Languagetree is not valid against the parse tree in tree-sitter, and is called fairly often

I attempted to mention this in the docstring of invalidate() https://github.com/neovim/neovim/pull/28747

DanilaMihailov committed:
test(treesitter): add tests for reloading parser

DanilaMihailov commented on test/functional/treesitter/parser_spec.lua:
```diff
@@ -959,4 +959,57 @@ int x = INT_MAX;
+      [[
+      vim.treesitter.get_parser(0, "lua")
+      vim.opt.rtp:append(...)
+      collectgarbage() -- without it keeps cached queries
```
@clason speaking about `vim.treesitter.query.get()`, I hit cache in tests, but never did when just running nvim, so I thought I misunderstood how caching suppose to work there.

I guess we need to invalidate query cache when getting new parser with `reload = true`?

clason replied:
Yes, indeed, as reloading the queries is what you want in the first place. Possibly it's enough for this purpose if we allow invalidating the cache explicitly (maybe for now with a private function prefixed by `_`).

DanilaMihailov replied:
I did not completely get what you mean by "invalidating the cache explicitly with a private function", could you elaborate, please?

Looked into `memoize` function and it is a closure over `cache`, and so cache is not accessible from outside. 

https://github.com/neovim/neovim/blob/4e5c633ed4871a948aff7338b793ac5f93484153/runtime/lua/vim/func/_memoize.lua#L41-L63

It can be converted to a table with `__call` that calls original function and some method like `_invalidate` attached, that clears cache (or specific key). And this would allow us to do something like this `vim.treesitter.query.get._invalidate_cache()`.

Or another option is to change what parameters used for cache, I tried this

```lua
M.get = function(lang, query_name)
  return getCachedQuery(lang, query_name, vim.o.rtp)
end
```

and `getCachedQuery` is original function, with one param for cache
```lua
local getCachedQuery = memoize('concat-3', function(lang, query_name, _rtp)
...
end)
```

DanilaMihailov replied:
changed `memoize` could look something like this
```lua
return function(hash, fn, strong)
  -- ...
  local cache = {}
  local invalidate_cache = function(...)
    if #{ ... } == 0 then
      cache = {}
    else
      cache[hash(...)] = nil
    end
  end

  local wrapped = function(_self, ...)
    -- ...
  end

  return setmetatable({ _invalidate_cache = invalidate_cache }, { __call = wrapped })
end
```

DanilaMihailov replied:
Settled for changing memozie function and added `treesitter.query._reset_cache` function, that calls _reset_cache internally, but allows simplier api for clearing all queries for lang: `query._reset_cache("lua")`.

Updated tests as well.

DanilaMihailov committed:
feat: add ability to reset query cache

DanilaMihailov committed:
fix: reset query cache when reloading parser

DanilaMihailov committed:
test: update parser reload tests

DanilaMihailov committed:
docs: run make doc

lewis6991 commented:
> There is no way to reload current parser for buffer. This is needed when
vim.opt.rtp changes and new paths have treesitter queries. For example
when loading plugins lazily.

Which plugins are we talking about? Is the lazy loading being done meaningfully or just blindly? And why can't the plugin be improved so it doesn't need to be lazy loaded (assuming it does).

It's still up in the air whether we want to support `rtp` changes, since there is so much state that needs to be reset and will get more difficult if we want to add more caches. I'm not sure if the added complexity is worth it for the 0.2ms startup time you get from avoiding a specific plugins `plugin/` dir.

DanilaMihailov commented:
> Which plugins are we talking about? Is the lazy loading being done meaningfully or just blindly? And why can't the plugin be improved so it doesn't need to be lazy loaded (assuming it does).

This one https://github.com/S1M0N38/love2d.nvim , I actually added the queries there and thats when I hit the problem with rtp changes.

> It's still up in the air whether we want to support rtp changes, since there is so much state that needs to be reset and will get more difficult if we want to add more caches. I'm not sure if the added complexity is worth it for the 0.2ms startup time you get from avoiding a specific plugins plugin/ dir.

In my case I load this plugin only in specific projects, so it is not about speed, but more about per project configuration. 

There is similar idea in another PR https://github.com/neovim/neovim/pull/27392 that may solve my problem.

lewis6991 commented:
love2d.nvim creates two user commands: https://github.com/S1M0N38/love2d.nvim/blob/main/plugin/love2d.lua

Is having this enabled only for specific projects really gaining you anything?

DanilaMihailov commented:
> love2d.nvim creates two user commands: https://github.com/S1M0N38/love2d.nvim/blob/main/plugin/love2d.lua
> 
> Is having this enabled only for specific projects really gaining you anything?

It adds love2d library to lua_ls path as well. But sure, this plugin can be loaded eagerly without a problem. I will have some irrelevant functions in autocomplete, but it is not a big deal.

Before creating PR I tried to solve this with available apis, but had to use private field
```lua
local parser = vim.treesitter.get_parser()
  parser._injection_query = vim.treesitter.query.get("lua", "injections")
  parser:invalidate(true)
  parser:parse(true)
```

Weirdly never hit cache with this call, so it worked, but using private field bothers me :)

DanilaMihailov commented:
> It's still up in the air whether we want to support rtp changes, since there is so much state that needs to be reset and will get more difficult if we want to add more caches.

The whole parser already getting dropped and new one created when executing `:edit` command for example

https://github.com/neovim/neovim/blob/6a264e08974bcb1b91f891eb65ef374f350d2827/runtime/lua/vim/treesitter.lua#L47-L52

https://github.com/neovim/neovim/blob/6a264e08974bcb1b91f891eb65ef374f350d2827/runtime/lua/vim/treesitter.lua#L106-L109

I am doing the same, when calling `vim.treesitter.get_parser(..., {reload = true})`, so no state from previous parser is kept or managed in any way. The only thing is I have to clear cache for queries, which does not seem like a big deal. But you saying there might be added more query files and then more cache is needed to be cleared?

DanilaMihailov commented:
If explicitly dropping query cache is a problem (or could become a problem in the future) how about just having an api to remove parser from a buffer?

```lua
function M.drop_parser(bufnr)
  parsers[bufnr]:destroy()
  parsers[bufnr] = nil
end
```

Or making `parsers` available outside of module? Like in `vim.treesitter.highliger.active[bufnr]`?

Or even adding parameter to `vim.treesitter.stop(bufnr, {drop = true})` that will delete parser?

This will enable me to write something like this (probably in a loop for all lua buffers)
```lua
vim.treesitter.stop(bufnr, {drop = true})
collectgarbage() -- maybe clear query cache
vim.treesitter.start(bufnr, "lua") -- create new parser with fresh queries
```

clason commented:
Remember that the highlighter is not the only "module" that is using treesitter parsers (which are shared!) and queries.

DanilaMihailov commented:
> Remember that the highlighter is not the only "module" that is using treesitter parsers (which are shared!) and queries.

Yeah, but they already have to handle buffer detaching 
https://github.com/neovim/neovim/blob/6a264e08974bcb1b91f891eb65ef374f350d2827/runtime/lua/vim/treesitter.lua#L47-L52

So calling `self:_on_detach()` should be enough to handle this. 

```lua
function M.drop_parser(bufnr)
  if parsers[bufnr] then
    parsers[bufnr]:_on_detach()
    parsers[bufnr] = nil
  end
end
```

Regarding queries, query cache already drops if no one is referencing given query. So by removing parser this way I try to create situation where no one references `queries/lua/injections.scm`, so it will be dropped from cache.

DanilaMihailov commented:
@lewis6991 please, let me know if this is a good idea. This is the only needed change. Not touching query cache. Just an ability to optionally destroy a parser. 

```lua
-- vim/treesitter.lua
function M.stop(bufnr, opts)
  opts = opts or {}
  -- ...
  if opts.destroy and parsers[bufnr] then
    parsers[bufnr]:_on_detach()
    parsers[bufnr]:destroy()
    parsers[bufnr] = nil
  end
  -- ...
end
```

justinmk commented:
> let me know if this is a good idea. This is the only needed change. Not touching query cache.

@DanilaMihailov that looks like a good start to me. Even if we can't get it perfect, we need to introduce an interface for this and start hacking away at an implementation. It's very annoying to develop parsers if I can't clear away the existing parser.

Related: https://github.com/neovim/neovim/pull/29530
